### PR TITLE
Fix SM90 DSA backward dS scaling precision

### PR DESCRIPTION
**Pull request by icavan**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I reviewed the Hard Rules in the `AGENTS.md` for each directory this PR touches and the changes comply.
- [ ] I added GitHub labels: one `cat-*`, one or more `area:*` / `op:*`, and one `orig-*`.

## Affected area

FE OSS kernels or CuTeDSL

## Summary

Scale SM90 DSA backward `dS` while it is still FP32, then narrow it once to the configured 16-bit MMA operand type.

This issue is specific to the SM90 implementation. SM100 already passes `scale_softmax` into `quantize`, which multiplies the FP32 fragment before converting it to the MMA operand dtype. This change aligns the SM90 cast order with the existing SM100 path; SM100 requires no change.

## Why

The previous SM90 path narrowed unscaled `dS` before multiplying by `softmax_scale`, then narrowed it again. This implementation detail differed from SM100, introduced an extra BF16 rounding, and could overflow a finite FP32 intermediate in the FP16 path. The affected `dS` feeds both dQ and the dK portion of dKV.

## Related issues

None.

## API and compatibility impact

No API change. Numerical results are corrected only for the SM90 DSA backward kernel. A 1000-iteration BF16 benchmark was unchanged within noise (2.224 ms before, 2.230 ms after).

## Testing

Run on an SM90 CUDA environment:

- Regression test against the parent commit: 2 failed as expected (BF16 rounding mismatch; FP16 intermediate overflow).
- Regression test with this change: 2 passed.
- `pytest test/python/fe_api/dsa/test_DSA_sparse_attention_backward.py -k sm90`: 15 passed.
- Full test file: 34 passed, 32 skipped.
- Pre-commit checks on both changed files passed.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved numerical accuracy for sparse attention backward computations on SM90 GPUs.
  * Corrected gradient scaling to preserve FP32 precision before reduced-precision processing, improving `dKV` results.

* **Tests**
  * Added regression coverage verifying accurate gradient calculations, including nonzero query-tail inputs and finite-result checks.

### CODE REVIEW

File: test/python/fe_api/dsa/test_DSA_sparse_attention_backward.py [hunk 1]
```diff
@@ -1118,6 +1118,70 @@ def test_DSA_sparse_attention_backward_sm100_accumulates_odo_in_fp32():
...
+    old_dq_value = (prematurely_narrowed_ds.float() * kv_scalar.float()).to(dtype).item()
+    assert expected_dq_value != old_dq_value
+
+    expected_dq = torch.zeros_like(q)
+    expected_dq[:, :, :head_dim_v] = expected_dq_value
+    assert torch.equal(result["dq"], expected_dq)
+    assert torch.isfinite(result["dkv"]).all()
+    assert torch.isfinite(result["d_sink"]).all()
+
+
```
[the remark this answers is not in the document]
Comment (icavan) [reply test/python/fe_api/dsa/test_DSA_sparse_attention_backward.py:1191]: Addressed in 6c2a7ff. Q is now nonzero only in the 512:576 tail while the KV tail remains zero, so scores and the uniform P construction are unchanged, but dKV[:, 512:576] is a pure dS^T @ Q term. The test checks the FP32-scaled reference against the premature-conversion result. On the parent kernel it reports BF16 8.8125 vs 8.75 and FP16 inf vs 26256; both cases pass with the fix. SM90 tests: 15 passed; full test file: 34 passed, 32 skipped.

**Comment by jiayus-nvidia** [comment]
@cudnn-ci-bot run oss

**Review by jiayus-nvidia — APPROVED** [review]
